pkg/scanners/scannerv4/nodescan_convert.go

```diff
@@ -0,0 +1,134 @@
+package scannerv4
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	v4 "github.com/stackrox/rox/generated/internalapi/scanner/v4"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/protocompat"
+	"github.com/stackrox/rox/pkg/utils"
+)
+
+var (
+	rhcosOSImageRegexp = regexp.MustCompile(`(Red Hat Enterprise Linux) (CoreOS) ([\d])([\d]+)`)
+)
+
+func ToNodeScan(r *v4.VulnerabilityReport, osImageRef string) *storage.NodeScan {
+	// TODO(ROX-26593): Instead of fixing labels here, add RHCOS DistributionScanner to ClairCore
+	fixedNotes := fixNotes(toStorageNotes(r.Notes), osImageRef)
+
+	return &storage.NodeScan{
+		ScanTime:        protocompat.TimestampNow(),
+		Components:      toStorageComponents(r),
+		Notes:           fixedNotes,
+		ScannerVersion:  storage.NodeScan_SCANNER_V4,
+		OperatingSystem: toOperatingSystem(osImageRef),
+	}
+}
+
+func toOperatingSystem(ref string) string {
+	r := rhcosOSImageRegexp.FindStringSubmatch(ref)
+	if len(r) != 5 {
+		return ""
+	}
+	return fmt.Sprintf("rhcos:%s.%s", r[3], r[4])
+}
+
+func toStorageComponents(r *v4.VulnerabilityReport) []*storage.EmbeddedNodeScanComponent {
+	packages := r.GetContents().GetPackages()
+	result := make([]*storage.EmbeddedNodeScanComponent, 0, len(packages))
+
+	for _, pkg := range packages {
+		vulns := getPackageVulns(pkg.GetId(), r)
+		result = append(result, createEmbeddedComponent(pkg, vulns))
+	}
+	return result
+}
+
+func getPackageVulns(packageID string, r *v4.VulnerabilityReport) []*storage.EmbeddedVulnerability {
+	vulns := make([]*storage.EmbeddedVulnerability, 0)
+	mapping, ok := r.GetPackageVulnerabilities()[packageID]
+	if !ok {
+		// No vulnerabilities for this package, skip
+		return vulns
+	}
+	for _, vulnID := range mapping.GetValues() {
+		vulnerability, ok := r.Vulnerabilities[vulnID]
+		if !ok {
+			log.Warnf("Unable to find vulnerability %s in report - skipping this vulnerability", vulnID)
+			continue
+		}
+		vulns = append(vulns, convertVulnerability(vulnerability))
+	}
```

Comment:
The image scan equiv has a check to see if a vuln id has already been processed for a component and ignores it if so. Any concern with that scenario here? (granted I believe on insertion into DB we may dedupe then as well, but that shouldn't be relied on).